files/default/ext_auth_files/pcluster_dcv_ext_auth.py

```diff
@@ -0,0 +1,362 @@
+        )
+
+
+class ThreadedHTTPServer(ThreadingMixIn, HTTPServer):
+    """Handle requests in a separate thread."""
+
+
+def set_server_logging(server):
+    log_file = open(LOG_FILE_PATH, "w")
+    os.chmod(LOG_FILE_PATH, 0o644)
+    server.log_file = log_file
+
+
+def run_server(port, certificate=None, key=None):
+    """
+    This class run the external authenticator server on localhost.
+
+    The external authenticator *must* be run as a separate user, and the script *must* be executed in the home folder of
+    the user. The user has to have in his home an empty folder, /.accepted_sessions and an executable file with the
+    setuid bit set, write_token.exe.
+
+    :param port: the port in which you want to start the server
+    :param certificate: the certificate to use if https
+    :param key: the key to use if https and it's not in certificate
+    """
+    server_class = ThreadedHTTPServer
+    handler_class = ExtAuthHandler
+    server_address = ("localhost", port)
+    httpd = server_class(server_address, handler_class)
+    # the following error is not real.
+    set_server_logging(httpd)
+    if certificate:
+        if key:
+            httpd.socket = ssl.wrap_socket(httpd.socket, certfile=certificate, keyfile=key, server_side=True)
+        else:
+            httpd.socket = ssl.wrap_socket(httpd.socket, certfile=certificate, server_side=True)
+    print(
+        "Starting DCV external authenticator {PROTOCOL} server on port {PORT}, use <Ctrl-C> to stop".format(
+            PROTOCOL="HTTPS" if certificate else "HTTP", PORT=port
+        )
+    )
+    httpd.serve_forever()
+
+
+def add_argument(parser, name, description, expected_type=str):
+    parser.add_argument(name, help=description, type=expected_type)
+
+
+def get_arguments():
+    parser = argparse.ArgumentParser(description="Execute the ParallelCluster External Authenticator")
+    add_argument(parser, "--port", "The port in which you want to start the server", expected_type=int)
+    add_argument(
+        parser, "--certificate", "The certificate to use for the external authenticator to run in HTTPS. Has to be .pem"
+    )
+    add_argument(parser, "--key", "The .key of the certificate, if not included in it")
+    return parser.parse_args()
+
+
+def generate_random_token(token_length):
+    """This function generate CSPRNG compliant random tokens."""
+    allowed_chars = "".join((string.ascii_letters, string.digits, "_", "-"))
+    max_int = len(allowed_chars) - 1
+    system_random = random.SystemRandom()
+    return "".join(allowed_chars[system_random.randint(0, max_int)] for _ in range(token_length))
+
+
+def generate_sha512_hash(*args):
+    """This function generates a sha512 """
+    hash_handler = hashlib.sha512()
+    for arg in args:
+        hash_handler.update(str(arg).encode("utf-8"))
+    return hash_handler.hexdigest()
+
+
+def main():
+    try:
+        args = get_arguments()
+        shutil.rmtree(USER_CREATED_FILE_DIR, ignore_errors=True)
```

Comment:
let's add a comment here to describe we are cleaning up the folder containing authorization files.